scripts/pre-push

```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+# A hook script to verify what is about to be pushed.  Called by "git
+# push" after it has checked the remote status, but before anything
+# has been pushed.  If this script exits with a non-zero status
+# nothing will be pushed.
+#
+# This hook is called with the following parameters:
+#
+# $1 -- Name of the remote to which the push is being done
+# $2 -- URL to which the push is being done
+#
+# If pushing without using a named remote those arguments will be equal.
+#
+# Information about the commits which are being pushed is supplied as
+# lines to the standard input in the form:
+#
+#   <local ref> <local sha1> <remote ref> <remote sha1>
+#
+# This script prevents pushing if the most recent commit does not pass
```

Comment:
Is this true? I think git hooks operate on the worktree (i.e. including uncommitted changes). The `--git` flag means "filepaths tracked by git" but not "contents of the last commit".

Reply:
You're right. I assumed a clean worktree, but it doesn't have to be the case. Fixing it...

Reply:
I am not sure what to do here without complicating it too much. I would just run either `hlint` or `hlint --git`; the latter seems like a better option as it more closely matches what we want to achieve here. First of all, I'm not sure what exactly the `--git` option does: does it "Run on files tracked by git" as in running it on committed changes only or does it run on files tracked by git, even though there might be unstaged changes to such files?

To make sure that hlint reports only issues with things that were committed, to use `hlint --git` one wouldd have to stash away the changes in a dirty working directory before linting.

Reply:
> or does it run on files tracked by git, even though there might be unstaged changes to such files?

This.

Reply:
It's still better than `hlint .` – this one will run on e.g. `./dist-newstyle`, which is completely improper.

Reply:
> one wouldd have to stash away the changes in a dirty working directory before linting.

Yes. Note that we don't have to add a note about Git hooks – I merely suggested it because if you find a hook helpful, it's nice to have it in the docs. If you change your mind about having a Git hook, I will not step up as a champion – I don't use hooks myself.

Reply:
Any thoughts, @neongreen ?

Reply:
> Yes. Note that we don't have to add a note about Git hooks – I merely suggested it because if you find a hook helpful, it's nice to have it in the docs. If you change your mind about having a Git hook, I will not step up as a champion – I don't use hooks myself.

For me `hlint --git` does the job. Is it OK to have it that way in the contribution guidelines (with a clarification note what exactly it runs on)?

Reply:
Sure.